File: exercicios_por_semana/semana_02_algebra_linear_i/01_vetores_retorno_risco/cpp/main.cpp
/*
Vetores - Retorno, risco e similaridade como operacoes vetoriais

Objetivo:
Usar produto interno, norma e projecao para comparar uma estrategia com o mercado.

Dados de partida:
- strategy = [0.012, -0.004, 0.006, 0.010, -0.008, 0.003, 0.007, -0.002]
- market   = [0.010, -0.006, 0.004, 0.011, -0.010, 0.002, 0.006, -0.003]
- risk_free_daily = 0.0002

O que fazer em C++:
1. guardar os retornos em vector<double>
2. implementar funcoes dot, norm_l2 e projection_scale
3. calcular cos(strategy_excess, market_excess)
4. calcular residual = strategy_excess - beta * market_excess
5. imprimir beta, norma do residual e cosine similarity

Saidas esperadas:
- beta/projecao da estrategia no mercado
- cosine similarity entre estrategia e mercado
- norma do residual que nao foi explicado pelo mercado
*/

#include <cmath>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>
#include <sstream>

struct ExerciseMetrics {
    double primary = 0.0;
    double secondary = 0.0;
};

double mean(const std::vector<double>& values) {
    double s = 0;
    for(double v: values)
        s +=v;

    return s / values.size();
}

std::vector<double> scale_vector(double beta, const std::vector<double>& values){
    std::vector<double> sv(values.size());
    for(auto i{0uz};i<values.size();i++)
        sv[i] = beta * values[i];

    return sv;
}

std::vector<double> subtract_scalar(const std::vector<double>& values, double scalar){
    std::vector<double> sv(values.size());
    for(auto i{0uz};i<values.size();i++)
        sv[i] = values[i] - scalar;

    return sv;
}


double dot(std::vector<double> u, std::vector<double> v){
    std::vector<double> mv(u.size());
    double d = 0;
    for(auto i{0uz};i<u.size();i++)
        mv[i] = u[i] * v[i];

    for(auto i{0uz};i<u.size();i++)
        d += mv[i];

    return d;
}



double norm_l2(const std::vector<double>& u){
    return std::sqrt(dot(u,u));
}


double cosine_similarity(const std::vector<double>& u, const std::vector<double>& v){
    return dot(u, v) / (norm_l2(u) * norm_l2(v));
}

double projection_beta(const std::vector<double>& strategy_excess, const std::vector<double>& market_excess) {
    return dot(strategy_excess, market_excess) / dot(market_excess, market_excess);
}


std::vector<double> subtract_vectors(const std::vector<double>& u, const std::vector<double>& v) {
    std::vector<double> s(u.size());

    for(auto i{0uz};i<u.size();i++)
        s[i] = u[i] - v[i];

    return s;
}

std::string format_vector(const std::vector<double>& values) {
    std::ostringstream out;

    out << std::fixed << std::setprecision(6);
    out << "[";

    for (std::size_t i{0}; i < values.size(); ++i) {
        if (i > 0) {
            out << ", ";
        }

        out << values[i];
    }

    out << "]";

    return out.str();
}

void print_vector(const std::string& name, const std::vector<double>& values){
    std::cout << "Vector " << name <<  ": " << format_vector(values);
}

// Escreva as implementacoes das funcoes declaradas acima.
// A estrutura de run_exercise() abaixo mostra onde cada funcao sera usada.

ExerciseMetrics run_exercise() {
    const std::vector<double> strategy{0.012, -0.004, 0.006, 0.010, -0.008, 0.003, 0.007, -0.002};
    const std::vector<double> market{0.010, -0.006, 0.004, 0.011, -0.010, 0.002, 0.006, -0.003};
    const double risk_free_daily = 0.0002;
    
    // 1. Retorno medio da estrategia.
    const double mean_strategy = mean(strategy);

    // 2. Retorno medio do mercado.
    const double mean_market = mean(market);

    // 3. Excesso de retorno da estrategia.
    const std::vector<double> strategy_excess = subtract_scalar(strategy, risk_free_daily);

    // 4. Excesso de retorno do mercado.
    const std::vector<double> market_excess = subtract_scalar(market, risk_free_daily);

    // 5. Beta da estrategia em relacao ao mercado.
    const double beta = projection_beta(strategy_excess, market_excess);

    // 6. Cosine similarity entre estrategia e mercado.
    const double similarity = cosine_similarity(strategy_excess, market_excess);

    // 7. Vetor projetado: parte explicada pelo mercado.
    const std::vector<double> explained = scale_vector(beta, market_excess);

    // 8. Vetor residual: parte nao explicada pelo mercado.
    const std::vector<double> residual = subtract_vectors(strategy_excess, explained);

    // 9. Norma do residual.
    const double norm_explained = norm_l2(explained);
    const double norm_residual = norm_l2(residual);

    // 10. Conclusao: alpha proprio ou exposicao ao mercado?
    const std::string conclusion =
        beta > 0.8 && similarity > 0.9 && norm_residual < norm_explained
            ? "forte exposicao ao mercado; pouco alpha proprio neste exemplo"
            : "ha residual relevante; investigar se existe alpha proprio";

    std::cout << std::fixed << std::setprecision(6);

    std::cout << "1. Retorno medio da estrategia\n";
    std::cout << "mean_strategy = " << mean_strategy << "\n\n";

    std::cout << "2. Retorno medio do mercado\n";
    std::cout << "mean_market = " << mean_market << "\n\n";

    std::cout << "3. Excesso de retorno da estrategia\n";
    print_vector("strategy_excess", strategy_excess);
    std::cout << "\n";

    std::cout << "4. Excesso de retorno do mercado\n";
    print_vector("market_excess", market_excess);
    std::cout << "\n";

    std::cout << "5. Beta da estrategia em relacao ao mercado\n";
    std::cout << "beta = " << beta << "\n\n";

    std::cout << "6. Cosine similarity entre estrategia e mercado\n";
    std::cout << "cosine_similarity = " << similarity << "\n\n";

    std::cout << "7. Vetor projetado: parte explicada pelo mercado\n";
    print_vector("explained", explained);
    std::cout << "norm_explained = " << norm_explained << "\n\n";

    std::cout << "8. Vetor residual: parte nao explicada pelo mercado\n";
    print_vector("residual", residual);
    std::cout << "\n";

    std::cout << "9. Norma do residual\n";
    std::cout << "norm_residual = " << norm_residual << "\n\n";

    std::cout << "10. Conclusao\n";
    std::cout << conclusion << "\n";

    return ExerciseMetrics{beta, norm_residual};
}

int main() {
    std::cout << "Vetores - Retorno, risco e similaridade como operacoes vetoriais\n";
    std::cout << "Siga os passos 1 a 10 do README.md.\n\n";
    run_exercise();
    return 0;
}

File: exercicios_por_semana/semana_02_algebra_linear_i/01_vetores_retorno_risco/python/main.py
"""
Vetores - Retorno, risco e similaridade como operacoes vetoriais

Este arquivo segue a mesma ordem da secao "Formulas por exercicio" do README.md.
Implemente de cima para baixo, passos 1 a 10.
"""

from __future__ import annotations

from math import sqrt


# Dados de partida do exercicio.
STRATEGY = [0.012, -0.004, 0.006, 0.010, -0.008, 0.003, 0.007, -0.002]
MARKET = [0.010, -0.006, 0.004, 0.011, -0.010, 0.002, 0.006, -0.003]
RISK_FREE_DAILY = 0.0002


def mean(values: list[float]) -> float:
    """Media simples: soma(values) / quantidade."""
    return sum(values) / len(values)


def subtract_scalar(values: list[float], scalar: float) -> list[float]:
    """Subtrai o mesmo numero de todos os elementos do vetor."""
    return [xi-scalar for xi in values]


def dot(u: list[float], v: list[float]) -> float:
    """Produto interno: u[0]*v[0] + ... + u[n-1]*v[n-1]."""
    return sum(ui * vi for ui, vi in zip(u, v))


def norm_l2(u: list[float]) -> float:
    """Norma L2: sqrt(dot(u, u))."""
    # TODO: use dot(u, u) e sqrt.
    return sqrt(dot(u,u))
    


def cosine_similarity(u: list[float], v: list[float]) -> float:
    """Similaridade de cosseno: dot(u, v) / (norm(u) * norm(v))."""
    return dot(u, v) / (norm_l2(u) * norm_l2(v))


def projection_beta(strategy_excess: list[float], market_excess: list[float]) -> float:
    """Beta da estrategia em relacao ao mercado."""
    return dot(strategy_excess, market_excess) / dot(market_excess, market_excess)

def scale_vector(beta: float, values: list[float]) -> list[float]:
    """Multiplica todos os elementos do vetor por beta."""
    # TODO: implemente beta * values[i] para cada i.
    return [x*beta for x in values]


def subtract_vectors(u: list[float], v: list[float]) -> list[float]:
    """Subtrai dois vetores elemento a elemento: u[i] - v[i]."""
    return [ui - vi for ui, vi in zip(u, v)]



def print_vector(name: str, values: list[float]) -> None:
    formatted = ", ".join(f"{value:.6f}" for value in values)
    print(f"{name} = [{formatted}]")


def run_exercise() -> dict[str, float]:
    """Executa os 10 passos do README.md."""

    # 1. Retorno medio da estrategia.
    mean_strategy = mean(STRATEGY)

    # 2. Retorno medio do mercado.
    mean_market = mean(MARKET)

    # 3. Excesso de retorno da estrategia.
    strategy_excess = subtract_scalar(STRATEGY, RISK_FREE_DAILY)

    # 4. Excesso de retorno do mercado.
    market_excess = subtract_scalar(MARKET, RISK_FREE_DAILY)

    # 5. Beta da estrategia em relacao ao mercado.
    beta = projection_beta(strategy_excess, market_excess)

    # 6. Cosine similarity entre estrategia e mercado.
    similarity = cosine_similarity(strategy_excess, market_excess)

    # 7. Vetor projetado: parte explicada pelo mercado.
    explained = scale_vector(beta, market_excess)

    # 8. Vetor residual: parte nao explicada pelo mercado.
    residual = subtract_vectors(strategy_excess, explained)

    # 9. Norma do residual.
    norm_explained = norm_l2(explained)
    norm_residual = norm_l2(residual)

    # 10. Conclusao: alpha proprio ou exposicao ao mercado?
    if beta > 0.8 and similarity > 0.9 and norm_residual < norm_explained:
        conclusion = "forte exposicao ao mercado; pouco alpha proprio neste exemplo"
    else:
        conclusion = "ha residual relevante; investigar se existe alpha proprio"

    print("1. Retorno medio da estrategia")
    print(f"mean_strategy = {mean_strategy:.6f}")
    print()

    print("2. Retorno medio do mercado")
    print(f"mean_market = {mean_market:.6f}")
    print()

    print("3. Excesso de retorno da estrategia")
    print_vector("strategy_excess", strategy_excess)
    print()

    print("4. Excesso de retorno do mercado")
    print_vector("market_excess", market_excess)
    print()

    print("5. Beta da estrategia em relacao ao mercado")
    print(f"beta = {beta:.6f}")
    print()

    print("6. Cosine similarity entre estrategia e mercado")
    print(f"cosine_similarity = {similarity:.6f}")
    print()

    print("7. Vetor projetado: parte explicada pelo mercado")
    print_vector("explained", explained)
    print(f"norm_explained = {norm_explained:.6f}")
    print()

    print("8. Vetor residual: parte nao explicada pelo mercado")
    print_vector("residual", residual)
    print()

    print("9. Norma do residual")
    print(f"norm_residual = {norm_residual:.6f}")
    print()

    print("10. Conclusao")
    print(conclusion)

    return {
        "mean_strategy": mean_strategy,
        "mean_market": mean_market,
        "beta": beta,
        "cosine_similarity": similarity,
        "norm_explained": norm_explained,
        "norm_residual": norm_residual,
    }


def main() -> None:
    print("Vetores - Retorno, risco e similaridade como operacoes vetoriais")
    print("Siga os passos 1 a 10 do README.md.")
    print()
    run_exercise()


if __name__ == "__main__":
    main()
